File: src/agents/world-keeper.ts
/**
 * World-Keeper — 世界观守护者
 * 维护设定一致性，提供世界背景细节
 */

import { BaseAgent } from './base-agent.js';
import type { AIProvider } from '../types/provider.js';
import type { AgentConfig, AgentRequest } from '../types/agent.js';

const DEFAULT_CONFIG: AgentConfig = {
  role: 'world-keeper',
  name: '世界观守护者',
  description: '当需要确认世界观设定、历史背景、地理信息、魔法规则或设定一致性时使用。',
  systemPrompt: `你是世界观守护者（World-Keeper），负责维护游戏世界的设定一致性。

你的核心职责：
1. 确保所有叙事内容符合世界观设定
2. 提供符合背景的细节描写（地理、文化、历史、魔法等）
3. 拒绝或修正违背设定的内容
4. 在不破坏设定的前提下，合理扩展世界细节

工作准则：
- 基于已有设定进行回答，不随意编造与设定矛盾的内容
- 当被问到未定义的细节时，给出合理推演并标注"推演"
- 回复以结构化信息为主，供 Narrator 整合使用
- 简洁明了，控制在 100-200 字

回复格式：
【设定确认】符合/不符合/需要补充
【细节】相关世界观细节
【建议】对叙事的建议（如有）`,
  temperature: 0.5,
};

export class WorldKeeper extends BaseAgent {
  private worldLore: string[] = [];

  constructor(provider: AIProvider, model?: string, configOverride?: Partial<AgentConfig>) {
    super({ ...DEFAULT_CONFIG, ...configOverride }, provider, model);
  }

  /** 加载世界观资料 */
  loadLore(lore: string[]): void {
    this.worldLore = lore;
  }

  protected buildSystemPrompt(request: AgentRequest): string {
    let prompt = this.config.systemPrompt;
    if (this.worldLore.length > 0) {
      prompt += `\n\n===== 世界设定资料 =====\n${this.worldLore.join('\n\n')}`;
    }
    return prompt;
  }
}

File: newgame/crates/core/src/agents/world_keeper.rs
use super::{AIProvider, AgentResponse, BaseAgent, Message, Role};
use anyhow::Result;
use std::sync::Arc;

pub struct WorldKeeperAgent {
    provider: Arc<dyn AIProvider>,
    temperature: f32,
    world_lore: String,
}

impl WorldKeeperAgent {
    pub fn new(provider: Arc<dyn AIProvider>) -> Self {
        Self {
            provider,
            temperature: 0.4,
            world_lore: String::new(),
        }
    }

    pub fn set_world_lore(&mut self, lore: &str) {
        self.world_lore = lore.to_string();
    }

    pub fn check_consistency(&self, content: &str) -> bool {
        if self.world_lore.is_empty() {
            return true;
        }
        let lower_lore = self.world_lore.to_lowercase();
        let lower_content = content.to_lowercase();
        let forbidden = ["魔法", "magick", "超能力", "superpower"];
        for word in &forbidden {
            if lower_lore.contains(word) && !lower_content.contains(word) {
                continue;
            }
        }
        true
    }
}

#[async_trait::async_trait]
impl BaseAgent for WorldKeeperAgent {
    fn name(&self) -> &str {
        "WorldKeeper"
    }

    fn system_prompt(&self) -> String {
        let lore_context = if self.world_lore.is_empty() {
            "No specific world lore has been set yet.".to_string()
        } else {
            format!("World Lore:\n{}", self.world_lore)
        };

        format!(
            r#"You are the World Keeper of an interactive fiction game. Your role is to maintain the consistency of the game world and enforce its established lore.

{}

Your responsibilities:
1. Verify that player actions and narrative events conform to the world's rules and lore.
2. Provide background details about the world when asked.
3. Reject or flag content that violates established world settings.
4. Ensure continuity in geography, history, magic systems, factions, and cultures.

When responding, first provide your internal reasoning inside 【观察】, 【分析】, and 【判定】 tags.
Then, output your final assessment or world information.
Keep responses concise and focused on lore consistency."#,
            lore_context
        )
    }

    async fn process_action(
        &self,
        context: &str,
        player_input: &str,
        history: &[Message],
    ) -> Result<AgentResponse> {
        let combined_context = if self.world_lore.is_empty() {
            context.to_string()
        } else {
            format!("{}\n\nWorld Lore:\n{}", context, self.world_lore)
        };

        let messages = self.build_messages(&combined_context, player_input, history);
        let raw_response = self
            .provider
            .generate_response(&messages, self.temperature)
            .await?;

        Ok(self.extract_chain_of_thought(&raw_response))
    }
}
